### PR TITLE
BobLight protocol

### PR DESCRIPTION
**Pull request by blazoncek**
Add (simplified) BobLight protocol implementation to WLED using usermod approach.

### CODE REVIEW

**Comment by softhack007** [comment]
👍 will review later this week.
Do you have some more infos about the protocol? timing, error codes, etc?

File: usermods/boblight/boblight.h [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+      int lightcount = 0;
+      int total = top+left+right+bottom;
+      int bcount;
+
+      if ( total > strip.getLengthTotal() ) {
+        DEBUG_PRINTLN(F("BobLight: Too many lights."));
+        return;
+      }
+
+      if ( bottom > 0 )
```
Comment (Aircoookie) [inline usermods/boblight/boblight.h:70]: Minor thing, but maybe align code style in this function, e.g. to `if (bottom > 0) {`.
Will test the usermod fully later :)

File: usermods/boblight/boblight.h [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+          lights[lightcount].vscan[0] = rcurrent;
+          lights[lightcount].vscan[1] = rtop;
+          lightcount+=1;
+          rcurrent = rtop;
+          rcount+=1;
+        }
+      }
+      
+            
+      if (bottom > 0) {
```
Comment (softhack007) [inline usermods/boblight/boblight.h:147]: same condition as line 69 🤔 is this part doing some after-processing, or maybe you merge it with the previous `if (bottom>0)` part? Again, a few comment would help the reader 😜

File: usermods/boblight/boblight.h [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+      bool en = enabled;
+      JsonObject um = root[FPSTR(_name)];
+      if (!um.isNull()) {
+        if (um[FPSTR(_enabled)].is<bool>()) {
+          en = um[FPSTR(_enabled)].as<bool>();
+        } else {
+          String str = um[FPSTR(_enabled)]; // checkbox -> off or on
+          en = (bool)(str!="off"); // off is guaranteed to be present
+        }
+        if (en != enabled && lights) enable(en);
```
Comment (softhack007) [inline usermods/boblight/boblight.h:287]: What will happen if I disable the usermod while boblights are receiving? Will it stop and exit Realtime Mode properly?

File: usermods/boblight/boblight.h (hunk 2 above)
Comment (blazoncek) [reply usermods/boblight/boblight.h:147]: There are 2 bottom "segments". Left and right.

File: usermods/boblight/boblight.h (hunk 3 above)
Comment (blazoncek) [reply usermods/boblight/boblight.h:287]: Need to test. Thanks.

File: usermods/boblight/boblight.h (hunk 2 above)
Comment (softhack007) [reply usermods/boblight/boblight.h:147]: 👍

**Review by softhack007 — APPROVED** [review]

**Comment by blazoncek** [comment]
Let's wait if @Aircoookie has something to say.

**Comment by Aircoookie** [comment]
Couldn't get a boblight daemon to work to test this, but reviewed your code and installed the usermod, found nothing to complain about :)

**Comment by blazoncek** [comment]
A Windows PC with Kodi installed should do.
But then you merged it. 😁

**Pull request merged**